diff --git a/tools/releasetools/add_img_to_target_files.py b/tools/releasetools/add_img_to_target_files.py
index babfc7dca..c618531fd 100644
--- a/tools/releasetools/add_img_to_target_files.py
+++ b/tools/releasetools/add_img_to_target_files.py
@@ -933,6 +933,26 @@ def AddImagesToTargetFiles(filename):
           if item not in vbmeta_system.split()]
       vbmeta_partitions.append("vbmeta_system")
 
+    vbmeta_system_ext = OPTIONS.info_dict.get("avb_vbmeta_system_ext", "").strip()
+    if vbmeta_system_ext:
+      banner("vbmeta_system_ext")
+      partitions["vbmeta_system_ext"] = AddVBMeta(
+          output_zip, partitions, "vbmeta_system_ext", vbmeta_system_ext.split())
+      vbmeta_partitions = [
+          item for item in vbmeta_partitions
+          if item not in vbmeta_system_ext.split()]
+      vbmeta_partitions.append("vbmeta_system_ext")
+
+    vbmeta_product = OPTIONS.info_dict.get("avb_vbmeta_product", "").strip()
+    if vbmeta_product:
+      banner("vbmeta_product")
+      partitions["vbmeta_product"] = AddVBMeta(
+          output_zip, partitions, "vbmeta_product", vbmeta_product.split())
+      vbmeta_partitions = [
+          item for item in vbmeta_partitions
+          if item not in vbmeta_product.split()]
+      vbmeta_partitions.append("vbmeta_product")
+
     vbmeta_vendor = OPTIONS.info_dict.get("avb_vbmeta_vendor", "").strip()
     if vbmeta_vendor:
       banner("vbmeta_vendor")
diff --git a/tools/releasetools/common.py b/tools/releasetools/common.py
index 6e3cd70a8..48ab27b5e 100644
--- a/tools/releasetools/common.py
+++ b/tools/releasetools/common.py
@@ -118,7 +118,7 @@ AVB_PARTITIONS = ('boot', 'dtbo', 'odm', 'product', 'pvmfw', 'recovery',
                   'vendor_dlkm', 'odm_dlkm')
 
 # Chained VBMeta partitions.
-AVB_VBMETA_PARTITIONS = ('vbmeta_system', 'vbmeta_vendor')
+AVB_VBMETA_PARTITIONS = ('vbmeta_system', 'vbmeta_system_ext', 'vbmeta_product', 'vbmeta_vendor')
 
 # Partitions that should have their care_map added to META/care_map.pb
 PARTITIONS_WITH_CARE_MAP = [
diff --git a/tools/releasetools/ota_from_target_files.py b/tools/releasetools/ota_from_target_files.py
index 56b784ddd..43db2a593 100755
--- a/tools/releasetools/ota_from_target_files.py
+++ b/tools/releasetools/ota_from_target_files.py
@@ -305,7 +305,7 @@ RETROFIT_DAP_UNZIP_PATTERN = ['OTA/super_*.img', AB_PARTITIONS]
 # 'system_other' and bootloader partitions.
 SECONDARY_PAYLOAD_SKIPPED_IMAGES = [
     'boot', 'dtbo', 'modem', 'odm', 'odm_dlkm', 'product', 'radio', 'recovery',
-    'system_ext', 'vbmeta', 'vbmeta_system', 'vbmeta_vendor', 'vendor',
+    'system_ext', 'vbmeta', 'vbmeta_system', 'vbmeta_system_ext', 'vbmeta_product', 'vbmeta_vendor', 'vendor',
     'vendor_boot']
 
 
diff --git a/tools/releasetools/sign_target_files_apks.py b/tools/releasetools/sign_target_files_apks.py
index 936ef888d..65b1e3aa2 100755
--- a/tools/releasetools/sign_target_files_apks.py
+++ b/tools/releasetools/sign_target_files_apks.py
@@ -213,6 +213,8 @@ AVB_FOOTER_ARGS_BY_PARTITION = {
     'vendor_dlkm': "avb_vendor_dlkm_add_hashtree_footer_args",
     'vbmeta': 'avb_vbmeta_args',
     'vbmeta_system': 'avb_vbmeta_system_args',
+    'vbmeta_system_ext': 'avb_vbmeta_system_ext_args',
+    'vbmeta_product': 'avb_vbmeta_product_args',
     'vbmeta_vendor': 'avb_vbmeta_vendor_args',
 }
 
@@ -1279,6 +1281,18 @@ def main(argv):
       OPTIONS.avb_algorithms['vbmeta_system'] = a
     elif o == "--avb_vbmeta_system_extra_args":
       OPTIONS.avb_extra_args['vbmeta_system'] = a
+    elif o == "--avb_vbmeta_system_ext_key":
+      OPTIONS.avb_keys['vbmeta_system_ext'] = a
+    elif o == "--avb_vbmeta_system_ext_algorithm":
+      OPTIONS.avb_algorithms['vbmeta_system_ext'] = a
+    elif o == "--avb_vbmeta_system_ext_extra_args":
+      OPTIONS.avb_extra_args['vbmeta_system_ext'] = a
+    elif o == "--avb_vbmeta_product_key":
+      OPTIONS.avb_keys['vbmeta_product'] = a
+    elif o == "--avb_vbmeta_product_algorithm":
+      OPTIONS.avb_algorithms['vbmeta_product'] = a
+    elif o == "--avb_vbmeta_product_extra_args":
+      OPTIONS.avb_extra_args['vbmeta_product'] = a
     elif o == "--avb_vbmeta_vendor_key":
       OPTIONS.avb_keys['vbmeta_vendor'] = a
     elif o == "--avb_vbmeta_vendor_algorithm":
@@ -1348,6 +1362,12 @@ def main(argv):
           "avb_vbmeta_system_algorithm=",
           "avb_vbmeta_system_key=",
           "avb_vbmeta_system_extra_args=",
+          "avb_vbmeta_system_ext_algorithm=",
+          "avb_vbmeta_system_ext_key=",
+          "avb_vbmeta_system_ext_extra_args=",
+          "avb_vbmeta_product_algorithm=",
+          "avb_vbmeta_product_key=",
+          "avb_vbmeta_product_extra_args=",
           "avb_vbmeta_vendor_algorithm=",
           "avb_vbmeta_vendor_key=",
           "avb_vbmeta_vendor_extra_args=",
